### DISCUSSION TITLE
长标题在页眉的显示问题

### DISCUSSION THREAD

**Discussion by RenBaixiao** [stone-zeng/fduthesis#264] [Q&A] [answered]
长标题在封面上需要两行进行显示，使用 `\\` 进行了手动换行，
但是如此一来，页眉显示的标题也会跟着变成两行了，很不美观且编译会报警告，
请问如何修改代码使页眉的标题总是一行呢？
感谢！

**Comment by stone-zeng** [comment]
这个我会修一下，让多行连成一行（#265）。但如果标题本来就很长的话可能还是会有溢出的情况，也有可能和右边的 chapter/section 重叠。

**Reply by RenBaixiao** [reply]
我的论文标题并没有超过学校字数限制，一行是没问题的，偶数页页眉是标题，奇数页页眉是章节，也不会有问题。
我先不手动断句了，实在不行两种情况都编译一遍，最后把手动换行的封面页拼过去

**Reply by stone-zeng** [reply]
话说学校 / 院系有规定页眉填写什么内容吗？（因为默认情况下模板并没有用到标题作为页眉，这样的话我就不修 #265 了）

**Comment by atxy-blip** [comment, marked as answer]
一种临时措施见 <https://tex.stackexchange.com/q/6862>，用 `\chaptermark` 和 `\sectionmark` 这两个命令即可覆盖页眉显示的标题：

```TeX
\chapter[目录标题]{原始标题}
\chaptermark{页眉标题}

\section[目录标题]{原始标题}
\sectionmark{页眉标题}
```

**Reply by RenBaixiao** [reply]
抱歉，我描述时有误，应该是整篇论文的标题在封面，会自动换行用两行显示，为了美观和合理断句，用"\\"进行手动换行，导致偶数页的页眉显示的标题也会两行，实际一行是完全没问题的
您的答案是处理章节标题和奇数页页眉的，请问文章标题有类似临时措施吗？
anyway，非常感谢

**Reply by note286** [reply]
我大概明白的你的意思，建议你上传几张截图更清晰，一图胜千言，避免误会。

**Reply by atxy-blip** [reply]
仅有封面的标题需要换行，那就在生成封面后删掉标题的换行符。在 `\begin{document}` 后面打一个补丁就可以：

```tex
...
\begin{document}
\ExplSyntaxOn
\tl_remove_all:Nn \l__fdu_info_title_tl { \\ }
\ExplSyntaxOff
...
```

**Reply by RenBaixiao** [reply]
试了一下这个做法，并不起作用，加了之后封面的标题也换行无效了

**Reply by atxy-blip** [reply]
在我的测试例里没有问题呢……

![image](https://user-images.githubusercontent.com/63113546/182018517-70fa9617-e974-4907-8fd1-ba0d95484c88.png)
![image](https://user-images.githubusercontent.com/63113546/182018544-6413b219-b0d9-4bbd-ae78-46e188840d0a.png)

**Reply by RenBaixiao** [reply]
又改了一下，可以了，非常感谢。
之前不行可能是因为我原先加页眉页脚的代码写在`\begin{document}`前面了，看到你的代码挪到后面去就可以了。
感谢！

**Reply by atxy-blip** [reply]
放在前面其实也是没有问题的，`\fancyhead` 并不会在定义页眉时直接展开变量中的值。不过既然能用就不必深究失效原因了。

```TeX
\documentclass{fduthesis}

\fdusetup{info/title={论文\\标题}}

\usepackage{zhlipsum}
\ExplSyntaxOn
\fancyhead[OR]{\l__fdu_info_title_tl}
\ExplSyntaxOff

\begin{document}
\ExplSyntaxOn
\tl_remove_all:Nn \l__fdu_info_title_tl { \\ }
\ExplSyntaxOff

\mainmatter\chapter{介绍}\zhlipsum[1-42]

\end{document}
```

**Comment by stone-zeng** [comment]
c1f9e3e24377e8822953f09a4e3f41582f19ae38 之后可以通过 `\l__fdu_info_title_sanitized_tl` 变量访问删除了 `\\` 和汉字间空格的标题：

```tex
\documentclass{fduthesis}

\fdusetup{info/title={论文\\标题}}

\usepackage{zhlipsum}
\ExplSyntaxOn
\fancyhead[OR]{\l__fdu_info_title_sanitized_tl}
\ExplSyntaxOff

\begin{document}

\mainmatter
\chapter{介绍}
\zhlipsum[1-42]

\end{document}
```
